pkg/component/logging/vali/vali.go

```diff
@@ -0,0 +1,1104 @@
+	ClusterType           string
+	IngressClass          string
+	ValiHost              string
+}
+
+type vali struct {
+	client         client.Client
+	namespace      string
+	secretsManager secretsmanager.Interface
+	k8sVersion     *semver.Version
+	values         Values
+}
+
+var _ component.Deployer = &vali{}
+
+// New creates a new instance of Vali deployer.
+func New(
+	client client.Client,
+	namespace string,
+	secretsManager secretsmanager.Interface,
+	k8sVersion *semver.Version,
+	values Values) (component.Deployer, error) {
+
+	switch {
+	case client == nil:
+		return nil, errors.New("client cannot be nil")
+	case len(namespace) == 0:
+		return nil, errors.New("namespace cannot be empty")
+	case len(values.ValiImage) == 0:
+		return nil, errors.New("vali image cannot be empty")
+	case len(values.CuratorImage) == 0:
+		return nil, errors.New("vali curator image cannot be empty")
+	case len(values.InitLargeDirImage) == 0:
+		return nil, errors.New("init-large-dir image cannot be empty")
+	case len(values.KubeRBACProxyImage) > 0 && len(values.TelegrafImage) == 0:
+		return nil, errors.New("when kube-RBAC-proxy image is present the telegraf image cannot be empty")
+	case len(values.TelegrafImage) > 0 && len(values.KubeRBACProxyImage) == 0:
+		return nil, errors.New("when telegraf image is present the kube-RBAC-proxy image cannot be empty")
+	case len(values.KubeRBACProxyImage) > 0 && len(values.TelegrafImage) > 0 && secretsManager == nil:
+		return nil, errors.New("secretManager cannot be nil")
+	case len(values.KubeRBACProxyImage) > 0 && len(values.TelegrafImage) > 0 && k8sVersion == nil:
+		return nil, errors.New("kubernetes version cannot be nil")
+	}
+
+	return &vali{
+		client:         client,
+		namespace:      namespace,
+		secretsManager: secretsManager,
+		k8sVersion:     k8sVersion,
+		values:         values,
+	}, nil
+}
+
+func (v *vali) Deploy(ctx context.Context) error {
+	var (
+		valiConfigMapName                string
+		telegrafConfigMapName            string
+		genericTokenKubeconfigSecretName string
+		valiResources                    *corev1.ResourceRequirements
+		registry                         = managedresources.NewRegistry(kubernetes.SeedScheme, kubernetes.SeedCodec, kubernetes.SeedSerializer)
+		resources                        = []client.Object{}
+	)
+	if v.values.HvpaEnabled && v.values.Replicas > 0 {
+		resources = append(resources, v.getHVPA())
+
+		//If HVPA is enabled then we have to set the resources as same the current ones to avoid redundant scale up.
+		currentResources, err := kubernetesutils.GetContainerResourcesInStatefulSet(ctx, v.client, kubernetesutils.Key(v.namespace, v1beta1constants.StatefulSetNameVali))
+		if err != nil {
+			return err
+		}
+		if len(currentResources) != 0 && currentResources[valiName] != nil {
+			valiResources = &corev1.ResourceRequirements{
+				Requests: currentResources[valiName].Requests,
+				Limits:   currentResources[valiName].Limits,
+			}
+		}
```

Comment:
This code can be dropped since this logic is now performed by `gardener-resource-manager` during reconciliation of the `ManagedResource`.